src/core/sock.h

```diff
@@ -78,6 +79,9 @@ struct nn_sock
     /*  Next endpoint ID to assign to a new endpoint. */
     int eid;
 
+    /*  Count of active holds against the socket. */
+    int holds;
```

Comment:
Seems legit. If the "real protection" you speak of could be decomposed into lower-level protections such as this, there's merit in considering it, but 1) that's too big a bite to take right now, and 2) it doesn't jump out as something that would be obviously successful (since it feels like the heavy-handedness of a blanket protection here is justified)